File: libnd4j/include/ops/declarable/generic/loss/distillation_kl_loss.cpp
/*
*  ******************************************************************************
*  *
*  *
*  * This program and the accompanying materials are made available under the
*  * terms of the Apache License, Version 2.0 which is available at
*  * https://www.apache.org/licenses/LICENSE-2.0.
*  *
*  * See the NOTICE file distributed with this work for additional
*  * information regarding copyright ownership.
*  * Unless required by applicable law or agreed to in writing, software
*  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
*  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
*  * License for the specific language governing permissions and limitations
*  * under the License.
*  *
*  * SPDX-License-Identifier: Apache-2.0
*  *****************************************************************************
*/
#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_distillation_kl_loss)
#include <ops/declarable/headers/loss.h>
#include <ops/declarable/helpers/distillation_kl_loss.h>
#include <helpers/ConstantShapeHelper.h>
namespace sd {
namespace ops {
CUSTOM_OP_IMPL(distillation_kl_loss, 2, 1, false, 0, 0) {
  auto studentLogits = INPUT_VARIABLE(0);  // [batch, classes]
  auto teacherLogits = INPUT_VARIABLE(1);  // [batch, classes]
  NDArray* hardLabels = block.width() > 2 ? INPUT_VARIABLE(2) : nullptr;  // optional [batch]
  auto output = OUTPUT_VARIABLE(0);
  double temperature = block.getTArguments()->size() > 0 ? T_ARG(0) : 4.0;
  double alpha = block.getTArguments()->size() > 1 ? T_ARG(1) : 0.5;
  helpers::distillationKLLoss(studentLogits, teacherLogits, hardLabels, output,
                               temperature, alpha, block.launchContext());
  return sd::Status::OK;
}
DECLARE_TYPES(distillation_kl_loss) {
  getOpDescriptor()->setAllowedInputTypes({ALL_FLOATS});
  getOpDescriptor()->setAllowedOutputTypes({ALL_FLOATS});
}
DECLARE_SHAPE_FN(distillation_kl_loss) {
  return SHAPELIST(ConstantShapeHelper::getInstance().scalarShapeInfo(
      sd::ArrayOptions::dataType(inputShape->at(0))));
}
//////////////////////////////////////////////////////////////////////////
CUSTOM_OP_IMPL(distillation_kl_loss_grad, 2, 2, false, 0, 0) {
  auto studentLogits = INPUT_VARIABLE(0);
  auto teacherLogits = INPUT_VARIABLE(1);
  NDArray* hardLabels = block.width() > 2 ? INPUT_VARIABLE(2) : nullptr;
  auto dLdStudent = OUTPUT_VARIABLE(0);
  auto dLdTeacher = OUTPUT_VARIABLE(1);
  double temperature = block.getTArguments()->size() > 0 ? T_ARG(0) : 4.0;
  double alpha = block.getTArguments()->size() > 1 ? T_ARG(1) : 0.5;
  helpers::distillationKLLossBp(studentLogits, teacherLogits, hardLabels,
                                  dLdStudent, dLdTeacher, temperature, alpha,
                                  block.launchContext());
  return sd::Status::OK;
}
DECLARE_TYPES(distillation_kl_loss_grad) {
  getOpDescriptor()->setAllowedInputTypes({ALL_FLOATS});
  getOpDescriptor()->setAllowedOutputTypes({ALL_FLOATS});
}
DECLARE_SHAPE_FN(distillation_kl_loss_grad) {
  return SHAPELIST(CONSTANT(inputShape->at(0)), CONSTANT(inputShape->at(1)));
}
}  // namespace ops
}  // namespace sd
#endif

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/loss/DistillationKLLoss.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.loss;

import lombok.Getter;
import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.nd4j.linalg.api.ops.impl.loss.bp.DistillationKLLossBp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Knowledge Distillation KL Divergence Loss.
 *
 * Computes the combined distillation loss:
 *   L = alpha * T^2 * KL(softmax(student/T) || softmax(teacher/T))
 *       + (1-alpha) * CE(student, hardLabels)
 *
 * Inputs:
 *   0: studentLogits [batch, classes]
 *   1: teacherLogits [batch, classes]
 *   2: hardLabels    [batch] (optional)
 *
 * Float args:
 *   0: temperature (default: 4.0)
 *   1: alpha (default: 0.5)
 *
 * Output:
 *   0: scalar loss value
 *
 * Adam Gibson
 */
@NoArgsConstructor
public class DistillationKLLoss extends DynamicCustomOp {

    @Getter private double temperature = 4.0;
    @Getter private double alpha = 0.5;

    public DistillationKLLoss(INDArray studentLogits, INDArray teacherLogits) {
        this(studentLogits, teacherLogits, 4.0, 0.5);
    }

    public DistillationKLLoss(INDArray studentLogits, INDArray teacherLogits, double temperature, double alpha) {
        super(new INDArray[]{studentLogits, teacherLogits}, null);
        this.temperature = temperature;
        this.alpha = alpha;
        addTArgument(temperature, alpha);
    }

    public DistillationKLLoss(INDArray studentLogits, INDArray teacherLogits, INDArray hardLabels,
                               double temperature, double alpha) {
        super(new INDArray[]{studentLogits, teacherLogits, hardLabels}, null);
        this.temperature = temperature;
        this.alpha = alpha;
        addTArgument(temperature, alpha);
    }

    public DistillationKLLoss(SameDiff sameDiff, SDVariable studentLogits, SDVariable teacherLogits) {
        this(sameDiff, studentLogits, teacherLogits, 4.0, 0.5);
    }

    public DistillationKLLoss(SameDiff sameDiff, SDVariable studentLogits, SDVariable teacherLogits,
                               double temperature, double alpha) {
        super(null, sameDiff, new SDVariable[]{studentLogits, teacherLogits}, false);
        this.temperature = temperature;
        this.alpha = alpha;
        addTArgument(temperature, alpha);
    }

    public DistillationKLLoss(SameDiff sameDiff, SDVariable studentLogits, SDVariable teacherLogits,
                               SDVariable hardLabels, double temperature, double alpha) {
        super(null, sameDiff, new SDVariable[]{studentLogits, teacherLogits, hardLabels}, false);
        this.temperature = temperature;
        this.alpha = alpha;
        addTArgument(temperature, alpha);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (tArguments.size() > 0) this.temperature = tArguments.get(0);
        if (tArguments.size() > 1) this.alpha = tArguments.get(1);
    }

    @Override
    public String opName() {
        return "distillation_kl_loss";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> grad) {
        if (args().length > 2) {
            return Arrays.asList(new DistillationKLLossBp(sameDiff, arg(0), arg(1), arg(2),
                temperature, alpha).outputVariables());
        }
        return Arrays.asList(new DistillationKLLossBp(sameDiff, arg(0), arg(1),
            temperature, alpha).outputVariables());
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        return Collections.singletonList(inputDataTypes.get(0));
    }

    @Override
    public int getNumOutputs() {
        return 1;
    }
}
